engine/verification/assigner/engine.go

```diff
@@ -87,12 +97,162 @@ func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
+	referenceBlockID := receipt.ExecutionResult.BlockID
+
+	log := log.With().
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).
+		Hex("reference_block_id", logging.ID(referenceBlockID)).
+		Hex("container_block_id", logging.ID(containerBlockID)).
+		Logger()
+
+	// verification node should be staked at the reference block id.
+	ok, err := e.stakedAtBlockID(referenceBlockID)
+	if err != nil {
+		log.Error().Err(err).Msg("could not verify stake of verification node for result at reference block id")
+		return
+	}
+	if !ok {
+		log.Debug().Msg("node is not staked at reference block id, receipt is discarded")
+		return
+	}
+
+	// chunk assignment
+	chunkList, err := e.chunkAssignments(context.Background(), &receipt.ExecutionResult)
+	if err != nil {
+		log.Error().Err(err).Msg("could not determine chunk assignment")
+		return
+	}
+	log.Info().
+		Int("total_assigned_chunks", len(chunkList)).
+		Msg("chunk assignment done")
+
+	// pushes assigned chunks to
+	e.processChunks(chunkList, resultID)
+}
+
+// processChunks receives a list of chunks all belong to the same execution result. It creates a chunk locator
+// for each of those chunks and stores the chunk locator in the chunks queue.
+//
+// Note that all chunks in the input chunk list are assume to be legitimately assigned to this verification node
+// (through the chunk assigner), and all belong to the same execution result.
+//
+// Deduplication of chunk locators is delegated to the chunks queue.
+func (e *Engine) processChunks(chunkList flow.ChunkList, resultID flow.Identifier) {
+	for _, chunk := range chunkList {
+		log := e.log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Hex("chunk_id", logging.ID(chunk.ID())).
+			Uint64("chunk_index", chunk.Index).Logger()
+
+		locator := &chunks.Locator{
+			ResultID: resultID,
+			Index:    chunk.Index,
+		}
+
+		// pushes chunk locator to the chunks queue
+		ok, err := e.chunksQueue.StoreChunkLocator(locator)
+		if err != nil {
+			log.Error().Err(err).Msg("could not push chunk locator to chunks queue")
+			continue
+		}
+		if !ok {
+			log.Debug().Msg("could not push duplicate chunk locator to chunks queue")
+			continue
+		}
+
+		// notifies chunk queue consumer of a new chunk
+		e.newChunkListener.Check()
+		log.Debug().Msg("chunk locator successfully pushed to chunks queue")
+	}
+}
+
+// stakedAtBlockID checks whether this instance of verification node has staked at specified block ID.
+// It returns true and nil if verification node is staked at referenced block ID, and returns false and nil otherwise.
+// It returns false and error if it could not extract the stake of node as a verification node at the specified block.
+func (e *Engine) stakedAtBlockID(blockID flow.Identifier) (bool, error) {
+	identity, err := protocol.IdentityAtBlockID(e.state, blockID, e.me.NodeID())
+	if err != nil {
+		return false, fmt.Errorf("could not extract staked identify of node at block %v: %w", blockID, err)
```

Comment:
@zhangchiqing but doesn't such an error reflect a serious internal or protocol level problem? I am concerned that a `false, nil` would silent the things down, and it can go completely unnoticed.

Reply:
So you are assuming we are always part of the node list at that block, which I'm afraid might not be true if we start the node from the epoch before its staked. So the node will finalized blocks that it hasn't staked yet. Right?

Reply:
right, but what if there was an internal error with the `state`, and the node is really staked, then it goes completely silent, right? I will apply the comment, but just wanted to make sure that we are not missing an error logging.

Reply:
fixed in https://github.com/onflow/flow-go/pull/427/commits/3b211176e05ee3d3284ea4c32a65a0cf68a24c6d